### PR TITLE
JAMES-2526 Bounce should not fail for "nullSender"

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/MailModifier.java [hunk 1]
```diff
@@ -108,7 +109,10 @@ public void setRemoteHost() {
 
     public void setRecipients(List<MailAddress> recipients) {
         if (!recipients.isEmpty()) {
-            mail.setRecipients(recipients);
+            mail.setRecipients(recipients
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/MailModifier.java]: please extract this stream filtering to a method named filterNullSenders

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/MailModifier.java [hunk 2]
```diff
@@ -142,7 +146,7 @@ public void setSubjectPrefix(Mail originalMail) throws MessagingException {
     public void setReplyTo(Optional<MailAddress> optionalReplyTo) throws MessagingException {
         if (optionalReplyTo.isPresent()) {
             MailAddress replyTo = optionalReplyTo.get();
-            if (replyTo.equals(SpecialAddress.NULL)) {
+            if (replyTo.isNullSender() || replyTo.equals(SpecialAddress.NULL)) {
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/MailModifier.java]: doesn't make sense to me. Can we have methods with relevant names for such action ? I don't know what `replyTo.equals(SpecialAddress.NULL)` is and why it's different from `isNullSender`

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java [hunk 3]
```diff
@@ -86,7 +87,9 @@ public void process(Mail originalMail) throws MessagingException {
 
             if (senderDomainIsValid(newMail)) {
                 // Send it off...
-                mailet.getMailetContext().sendMail(newMail);
+                if (!newMail.getRecipients().isEmpty()) {
+                    mailet.getMailetContext().sendMail(newMail);
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java]: why it's not handled by `sendMail` ? I mean, is there any legitimate use of `sendMail` without any recipient ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/util/SpecialAddressesUtils.java [hunk 4]
```diff
@@ -186,6 +190,7 @@ public SpecialAddressesUtils(RedirectNotify genericMailet) {
             case SENDER:
             case REVERSE_PATH:
                 return Optional.ofNullable(mail.getSender())
+                    .filter(address -> !address.isNullSender())
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/util/SpecialAddressesUtils.java]: well, it looks like we repeat ourselves. What about `mail.getNonNullSenders()` ? why don't we have to check `replyTo.equals(SpecialAddress.NULL)` this time ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/MailModifier.java (hunk 2 above)
Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/MailModifier.java]: Well, I must confess my understanding of the entire `redirect` package is not good at all.

It breaks everywhere as soon as you do a single change, and in random places... I don't even understand what these special addresses are...

So basically I ask myself the same question than you...

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java (hunk 3 above)
Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java]: I did have the same thought than you but I did not want to put too much logic in `FakeMailContext`

After thinking about it, *I considered preventing bad behaviours in the first place was desirable*

File: server/mailet/mailets/src/main/java/org/apache/james/transport/util/SpecialAddressesUtils.java (hunk 4 above)
Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/util/SpecialAddressesUtils.java]: Again, I don't understand that code... Just did the changes where needed to get :green_apple:  tests

**Pull request closed without merging**